src/typechecker.c

```diff
@@ -890,7 +889,35 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
     } break;
 
     /// Make sure a cast is even possible.
-    case NODE_CAST: TODO();
+    case NODE_CAST: {
+      Type *t_to = expr->type;
+      if (type_is_incomplete(t_to))
+        ERR(t_to->source_location, "Can not cast to incomplete type %T", t_to);
+
+      if (!typecheck_expression(ast, expr->cast.value))
+        return false;
+
+      Type *t_from = expr->cast.value->type;
+
+      // TODO: What do we do when from and to types are equal?
```

Comment:
Nothing ;Þ

Reply:
As in, the value of the cast is just the value of the expression being cast. I think it might also make sense to just outright remove the cast expression from the AST and replace it w/ the expression being cast, however, AST node removal is a bit scuffed atm and I’m not sure it would always work; we’d have to make sure never to keep around references to nodes from before calling `typecheck_expression()` on them, e.g. this
```c
Node *foo = parent_node->some_child_node;
if (!typecheck_expression(ast, foo)) ...
/// foo may be invalid here if it was replaced
```
doesn’t work. Instead we’d have to make sure to always do
```c
if (!typecheck_expression(ast, parent_node->some_child_node)) ...
Node *foo = parent_node->some_child_node;
```

Reply:
However, since we’re already replacing function references with the actual functions, we should be doing this already anyway, so I think we just need to go through sema and make sure we don’t have any dangling pointers to removed nodes, and then we *should* just be able to do perform this replacement here too.

Reply:
Does a cast not have to create a temporary in this case?

Reply:
Not if the types are equal. Casting e.g. an `integer` to an `integer` is a no-op. The value being cast is already an `integer` so we can just reuse it.